core/src/test/java/io/grpc/internal/MetricRecorderImplTest.java

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * Unit test for {@link MetricRecorderImpl}.
+ */
+@RunWith(JUnit4.class)
+public class MetricRecorderImplTest {
+
+  private static final String DESCRIPTION = "description";
+  private static final String UNIT = "unit";
+  private static final boolean ENABLED = true;
+  private static final ImmutableList<String> REQUIRED_LABEL_KEYS = ImmutableList.of("KEY1", "KEY2");
+  private static final ImmutableList<String> OPTIONAL_LABEL_KEYS = ImmutableList.of(
+      "OPTIONAL_KEY_1");
+  private static final ImmutableList<String> REQUIRED_LABEL_VALUES = ImmutableList.of("VALUE1",
+      "VALUE2");
+  private static final ImmutableList<String> OPTIONAL_LABEL_VALUES = ImmutableList.of(
+      "OPTIONAL_VALUE_1");
+  private static final DoubleCounterMetricInstrument DOUBLE_COUNTER_INSTRUMENT =
+      new DoubleCounterMetricInstrument(0, "counter0", DESCRIPTION, UNIT, REQUIRED_LABEL_KEYS,
+          OPTIONAL_LABEL_KEYS, ENABLED);
+  private static final LongCounterMetricInstrument LONG_COUNTER_INSTRUMENT =
+      new LongCounterMetricInstrument(1, "counter1", DESCRIPTION, UNIT, REQUIRED_LABEL_KEYS,
+          OPTIONAL_LABEL_KEYS, ENABLED);
+  private static final DoubleHistogramMetricInstrument DOUBLE_HISTOGRAM_INSTRUMENT =
+      new DoubleHistogramMetricInstrument(2, "histogram1", DESCRIPTION, UNIT,
+          Collections.emptyList(), REQUIRED_LABEL_KEYS, OPTIONAL_LABEL_KEYS, ENABLED);
+  private static final LongHistogramMetricInstrument LONG_HISTOGRAM_INSTRUMENT =
+      new LongHistogramMetricInstrument(3, "histogram2", DESCRIPTION, UNIT,
+          Collections.emptyList(), REQUIRED_LABEL_KEYS, OPTIONAL_LABEL_KEYS, ENABLED);
+  private final List<MetricInstrument> instruments = Arrays.asList(DOUBLE_COUNTER_INSTRUMENT,
+      LONG_COUNTER_INSTRUMENT, DOUBLE_HISTOGRAM_INSTRUMENT, LONG_HISTOGRAM_INSTRUMENT);
+  private final Set<String> metricNames = Stream.of("counter1", "counter2", "histogram1",
+          "histogram2")
+      .collect(Collectors.toCollection(HashSet::new));
+  private MetricSink mockSink = mock(MetricSink.class);
+  private List<MetricSink> sinks = Arrays.asList(mockSink, mockSink);
+  private MetricInstrumentRegistry registry;
+  private MetricRecorder recorder;
+  private ArgumentCaptor<DoubleCounterMetricInstrument> doubleCounterInstrumentCaptor =
+      ArgumentCaptor.forClass(DoubleCounterMetricInstrument.class);
+  private ArgumentCaptor<LongCounterMetricInstrument> longCounterInstrumentCaptor =
+      ArgumentCaptor.forClass(LongCounterMetricInstrument.class);
+  private ArgumentCaptor<DoubleHistogramMetricInstrument> doubleHistogramInstrumentCaptor =
+      ArgumentCaptor.forClass(DoubleHistogramMetricInstrument.class);
+  private ArgumentCaptor<LongHistogramMetricInstrument> longHistogramInstrumentCaptor =
+      ArgumentCaptor.forClass(LongHistogramMetricInstrument.class);
+
+
+  @Before
+  public void setUp() {
+    registry = MetricInstrumentRegistryAccessor.createMetricInstrumentRegistry(instruments,
```

Comment:
Can't this use the zero-arg constructor more like:

```java
private MetricInstrumentRegistry registry = MetricInstrumentRegistryAccessor.createMetricInstrumentRegistry();
private final DoubleCounterMetricInstrument doubleCounterInstrument =
    registry.registerDoubleCounter("counter0", DESCRIPTION, UNIT, REQUIRED_LABEL_KEYS,
        OPTIONAL_LABEL_KEYS, ENABLED);
```

That seems easier and allows the registry to guarantee its internal state is accurate. It'd also allow you to get rid of the two-arg constructor, which is a hacky constructor.

Reply:
Updated to use zero-arg constructor